Framework/Algorithms/src/Q1DWeighted.cpp

```diff
@@ -207,14 +214,23 @@ void Q1DWeighted::exec() {
       if (spectrumInfo.isMonitor(i) || spectrumInfo.isMasked(i))
         continue;
 
+      // Skip the masked bins
+      if (inputWS->hasMaskedBins(i)) {
```

Comment:
You are right that it is very slow and redundant for the masked bins. I was originally planning to reorganise the code in a separate issue (see #24701), which was promising orders of magnitude of speed up. Until this morning that I realised that there is a reason why it loops first over lambda. So I think I will actually need to implement some caching of geometry and bin masking information, since they do not depend on the wavelength bin. Marking In Progress meanwhile.